luarules/gadgets/unit_collision_damage_behavior.lua

```diff
@@ -18,52 +18,67 @@ local fallDamageMagnificationFactor = 14 --the multiplier by which default engin
+local spSetUnitVelocity = Spring.SetUnitVelocity
+local mathMin = math.min
+local mathMax = math.max
+local mathCeil = math.ceil
 
 local fallDamageMultipliers = {}
 local unitsMaxImpulse = {}
 local weaponDefIDImpulses = {}
-local overImpulseCooldowns = {}
-local transportingUnits = {}
+local transportedUnits = {}
 local unitMasses = {}
-
-local currentModulusFrame = 0
+local weaponDefIgnored = {}
+local unitInertiaCheckFlags = {} -- track inertia via velocity (units have constant mass)
+local gameFrame = 0
+local velocityWatchDuration = 3
 
 for unitDefID, unitDef in ipairs(UnitDefs) do
 	local fallDamageMultiplier = unitDef.customParams.fall_damage_multiplier or 1.0
-	fallDamageMultipliers[unitDefID] = fallDamageMultiplier*fallDamageMagnificationFactor
-	unitsMaxImpulse[unitDefID] = unitDef.mass/maxImpulseThreshold
+	fallDamageMultipliers[unitDefID] = fallDamageMultiplier * fallDamageMagnificationFactor
+	unitsMaxImpulse[unitDefID] = unitDef.mass / maxImpulseProportion
 	unitMasses[unitDefID] = unitDef.mass
 end
 
+for name, weaponDefID in pairs(Game.envDamageTypes or {}) do
+	weaponDefIgnored[weaponDefID] = true
+end
+
 for weaponDefID, wDef in ipairs(WeaponDefs) do
 	if wDef.damages and wDef.damages.impulseBoost and wDef.damages.impulseFactor then
 		weaponDefIDImpulses[weaponDefID] = {impulseboost = wDef.damages.impulseBoost, impulsefactor = wDef.damages.impulseFactor}
 	end
-end
 
-local function calculateImpulseData(unitDefID, damage, weaponDefID)
-	local impulse = (damage + weaponDefIDImpulses[weaponDefID].impulseboost) * weaponDefIDImpulses[weaponDefID].impulsefactor
-	local maxImpulse = unitsMaxImpulse[unitDefID]
-	local impulseMultiplier = mathmin(maxImpulse/impulse, 1)
-	return impulse, impulseMultiplier
-end
-
-local function isUnitBeingTransported(unitID)
-	for transportID, transportedID in pairs(transportingUnits) do
-		if transportingUnits[transportID][unitID] then
-			return true
+	
+	local function maxDamage(damages)
+		local damage = damages[0]
+		for i = 1, #damages do
+			damage = mathMax(damages[i], damage)
 		end
+		return damage
+	end
+	
+	--generate list of exempted weapons to improve performance
+	if wDef.damages and wDef.damages.impulseFactor == 0 or
+		(wDef.damages.impulseFactor < 0.15 and wDef.damages.impulseBoost < maxDamage(wDef.damages) * 0.2) then
+		weaponDefIgnored[weaponDefID] = true
 	end
-	return false
 end
 
-local function calculateMassToHealthRatioMultiplier(unitID, unitDefID)
+local function impulseData(unitDefID, weaponDefID, damage)
+	local impulseBoost = weaponDefIDImpulses[weaponDefID].impulseboost or 0
+	local impulseFactor = weaponDefIDImpulses[weaponDefID].impulsefactor or 1
+	local impulse = (damage + impulseBoost) * impulseFactor
+	local maxImpulse = unitsMaxImpulse[unitDefID]
+	local impulseMultiplier = mathCeil(mathMin(maxImpulse/impulse, 1) * 1000000) / 1000000 --we round to prevent the loss of impulse due to sheered tiny, tiny values being returned in UnitPreDamaged
```

Comment:
A more in-depth discussion of this line is needed

Reply:
Before rounding change, impulse would sometimes be too low
After rounding change, seemed more consistent.

Reply:
and now as I return to this part of the gadget, it's fine without rounding. I'm just gonna choose to believe it